### PR TITLE
Add asciiflow diagram to the csi-proxy

### PR DESCRIPTION
**Pull request by jayunit100**
Adding a diagram ( adopted from @ddebroy 's slack message and conversations with @marosset yesterday) Also added example files to look at to make it easier to grok the way its used by looking at SMB/Azure File implementations

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,9 +1,39 @@
 # CSI Proxy
 
-CSI Proxy is a binary that exposes a set of gRPC APIs around storage operations
-over named pipes in Windows. A container, such as CSI node plugins, can mount
-the named pipes depending on operations it wants to exercise on the host and
-invoke the APIs.
+CSI Proxy (which might be more aptly named "csi-node-proxy") is a binary that exposes a set of gRPC APIs around local storage operations for nodes in Windows. A container, such as CSI node plugins, can mount the named pipes depending on operations it wants to exercise on the host and
+invoke the APIs.  This allows a storage plugin to run as if were a CSI plugin on linux which had the ability to do "priveleged" actions on the windows host OS.
```
Comment (marosset) [inline README.md:4]:
```suggestion
invoke the APIs.  This allows a storage plugin to run as if were a CSI plugin on linux which have the ability to do "priveleged" actions on the windows host OS.
```

File: README.md [hunk 2]
```diff
@@ -1,9 +1,39 @@
...
+| csi.sock            |                               |
+|                     v                               v
+|                 +---+-------------------------------+------+
+|                 |                                          |
++---------------->|         Host Os (Windows)                |
+                  +------------------------------------------++
+```
+In the above diagram, there are 3 communication channels being utilized for CSI proxy:
+
+- `csi.sock` : This lives at /var/lib/kubelet/plugins/csi-plugin/csi.sock, and is used for the host OS to communicate with the CSI node plugin
```
Comment (ddebroy) [inline README.md:34]: Typically the path names in Windows (assuming OS installed in C:\ in host) would be `C:\var\lib\kubelet`. You may also want to add the qualifier "This usually lives ..." as the installation paths can be customized by the distribution of Kubernetes.

File: README.md [hunk 3]
```diff
@@ -1,9 +1,39 @@
...
+- `csi-plugin-reg.sock` : This lives here, /var/lib/kubelet/plugins_registry/csi-plugin-reg.sock, and is used for the kubelet itself to get metadata as needed from the csi plugin  
+- `/pipe` : /pipe/csi-proxy-api-group-v-... : This pipe lives at \pipe\csi-proxy-api-group-v-..., upon receipt of specific requests from the Node plugin, it carries out actions on the node plugins behalf.  Since it doesn't execute arbitrary commands on the host, its more secure then simply being a black-box proxy to run host commands.
```
Comment (ddebroy) [inline README.md:36]: Please use `\\.\pipe\csi-proxy-...` for named pipe path prefix.

**Review by ddebroy — COMMENTED** [review]
Thanks for the updates and the ascii diagram! Some comments below.

File: README.md [hunk 4]
```diff
@@ -1,9 +1,39 @@
...
+                                              +------------------+
++---------------------------------------------|                  |
+|                 +----------+                |                  |
+|                 |csi node  |                |   CSI Node Plugin|
+|                 |driver    +---------------->                  |
+|                 |registrar |                |                  |
+| NodeStageVolume ----^------+                ----------+--------+
+| NodeUnstageVol      |                                 |  Disk APIs
+| NodePublishVol      |GetNodeInfo                      |  Volume APIs
+| NodeResizeVol       |NodeRegistrationStatus    /pipe  |  SMB APIs
```
[the remark this answers is not in the document]
Comment (jayunit100) [reply README.md:16]: ran out of space :), ok to just reference it at the bottom?

File: README.md (hunk 3 above)
Comment (jayunit100) [reply README.md:36]: ok will fix this shortly.

File: README.md [hunk 5]
```diff
@@ -1,9 +1,39 @@
 # CSI Proxy
 
-CSI Proxy is a binary that exposes a set of gRPC APIs around storage operations
-over named pipes in Windows. A container, such as CSI node plugins, can mount
-the named pipes depending on operations it wants to exercise on the host and
-invoke the APIs.
+CSI Proxy (which might be more aptly named "csi-node-proxy") is a binary that exposes a set of gRPC APIs around local storage operations for nodes in Windows. A container, such as CSI node plugins, can mount the named pipes depending on operations it wants to exercise on the host and
```
[the remark this answers is not in the document]
Comment (jayunit100) [reply README.md:3]: fixed

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (jayunit100) [reply README.md:16]: done

File: README.md [hunk 6]
```diff
@@ -1,9 +1,39 @@
...
++---------------------------------------------|                  |
+|                 +----------+                |                  |
+|                 |csi node  |                |   CSI Node Plugin|
+|                 |driver    +---------------->                  |
+|                 |registrar |                |                  |
+| NodeStageVolume ----^------+                ----------+--------+
+| NodeUnstageVol      |                                 |  Disk APIs
+| NodePublishVol      |GetNodeInfo                      |  Volume APIs
+| NodeResizeVol       |NodeRegistrationStatus    /pipe  |  SMB APIs
+| NodeGetInfo         |                                 |  FileSystem/System APIs
```
[the remark this answers is not in the document]
Comment (jayunit100) [reply README.md:17]: done

File: README.md (hunk 2 above)
Comment (jayunit100) [reply README.md:34]: done

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (jayunit100) [reply README.md]: done

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (jayunit100) [reply README.md]: fixed

**Comment by jayunit100** [comment]
thanks @ddebroy PTAL

**Comment by jayunit100** [comment]
@marosset @PatickLang all set on this one ?

**Comment by marosset** [comment]
@andyzhangx @jingxu97 can one of you talk a look?

**Review by andyzhangx — APPROVED** [review]
/lgtm
/approve

File: README.md [hunk 7]
```diff
@@ -1,9 +1,40 @@
...
+|                 +---+-------------------------------+------+
+|                 |                                          |
++---------------->|         Host Os (Windows)                |
+                  +------------------------------------------++
+```
+In the above diagram, there are 3 communication channels being utilized for CSI proxy:
+
+- `csi.sock` : This usually lives at c:/var/lib/kubelet/plugins/csi-plugin/csi.sock, and is used for the host OS to communicate with the CSI node plugin
+- `csi-plugin-reg.sock` : This lives here, c:/var/lib/kubelet/plugins_registry/csi-plugin-reg.sock, and is used for the kubelet itself to get metadata as needed from the csi plugin .
+- `\\.\pipe\csi-proxy` : This will live in a pipe, i.e. at \\.\pipe/csi-proxy-api-group-v-... : upon receipt of specific requests from the Node plugin, it carries out actions on the node plugins behalf.  Since it doesn't execute arbitrary commands on the host, its more secure then simply being a black-box proxy to run host commands.
```
Comment (jingxu97) [inline README.md:37]: \\.\pipe/csi-proxy-api-group-v- --> \\.\pipe\csi-proxy-api-group-v-  ?

**Pull request merged**